Add an ssh tunnel option to the /proxy endpoint

cjcullen opened the pull request:
Key generation and propagation is handled on gce.
Tunnels are recycled after 5 minutes.
The ssh-proxy functionality is not enabled by default for the gce provider.

I can squash commits when this gets ready to merge.

@brendandburns

cjcullen committed:
Mount cloud-config files for cloudproviders in kube-apiserver & kube-controllermanager.

brendandburns committed:
Add tunnelling to the proxy API handler.

brendandburns committed:
Initial proxy tunnelling.

brendandburns committed:
Add key generation.

cjcullen committed:
Specify sshUser, sshKeyfile in kube-apiserver manifest.
Trim space on ssh key so GCE doesn't treat it as 2 lines.
A couple other minor fixes.

brendandburns committed:
Add the SSHTunnel transport to the kubelet client.

cjcullen committed:
Fix transport creation logic.

Refactor loadTunnels to allow one path for load, another for refresh.
Make SSHTunnelList.Close sleep for a minute before actually closing each tunnel.

cjcullen committed:
Some refactoring. Only selectively use ssh proxy.

Add NetworkName to gce.Config.
Add locking to uses of master.tunnels.

quinton-hoole commented:
This is a lot of new code.  What issue does it address?  Is it v1.0?

alex-mohr commented:
I think part of #3168 ?

cjcullen commented:
This addresses #3168 for the /proxy case (we won't need kubelet listening on the public internet for proxy).

brendandburns commented on pkg/util/ssh.go:
```diff
@@ -172,3 +153,95 @@ func MakePrivateKeySigner(key string) (ssh.Signer, error) {
+	return nil
+}
+
+func (l SSHTunnelList) Close() {
```
Document that this is async, and tunnels won't be closed when the method returns.

brendandburns commented on pkg/util/util.go:
```diff
@@ -508,3 +508,16 @@ func GetClient(req *http.Request) string {
 }
+
+func FileExists(filename string) (bool, error) {
+	file, err := os.Open(filename)
```
in a different context, it was pointed out to me that you could use ```os.Stat``` here and save yourself having to close the file.

brendandburns commented:
Generally LGTM, needs a rebase (and tests, if you felt like writing them...)

cjcullen committed:
Make sshproxy use a hostmount on master PD (don't spam sshKeys on upgrade/reboot).
Add comment describing what SSHTunnelList.Close() does.
Simplify util.FileExists.

cjcullen committed:
Change sshproxy to poll registry for nodes every 10 seconds (reduces window where closed tunnels from scaling down may exist).

kprobst commented:
Mostly LGTM, a couple of nits.

cjcullen commented:
Note: this is not enabled by default. On GCE, it will only be enabled if grains.proxy_ssh_user is set.

quinton-hoole commented:
@cjcullen and @brendanburns are working on the missing tests, which can come in a separate PR. This PR LGTM.
